File: src/bytecanary/dfa.py
"""
UTF-8 DFA state machine and dual scoring for byte-level evaluation.

Provides V_partial, V_binary_strict, V_binary_soft scoring.
"""

from dataclasses import dataclass, field
from enum import Enum, auto
from typing import List


class UTF8State(Enum):
    START = auto()
    EXPECT_1 = auto()
    EXPECT_2 = auto()
    EXPECT_3 = auto()
    EXPECT_2_AFTER_1 = auto()
    EXPECT_3_AFTER_1 = auto()
    EXPECT_3_AFTER_2 = auto()
    INVALID = auto()


STATE_PROGRESS = {
    UTF8State.START: 1.0,
    UTF8State.EXPECT_1: 0.5,
    UTF8State.EXPECT_2: 0.333,
    UTF8State.EXPECT_2_AFTER_1: 0.667,
    UTF8State.EXPECT_3: 0.25,
    UTF8State.EXPECT_3_AFTER_1: 0.5,
    UTF8State.EXPECT_3_AFTER_2: 0.75,
    UTF8State.INVALID: 0.0,
}


@dataclass
class UTF8Analysis:
    valid_chars: int
    invalid_bytes: int
    total_bytes: int
    final_state: UTF8State
    pending_bytes: int
    expected_total: int
    char_boundaries: List[int] = field(default_factory=list)

    @property
    def is_complete(self) -> bool:
        return self.final_state == UTF8State.START

    @property
    def is_valid_prefix(self) -> bool:
        return self.final_state != UTF8State.INVALID and self.invalid_bytes == 0

    @property
    def incomplete_progress(self) -> float:
        return STATE_PROGRESS.get(self.final_state, 0.0)


class UTF8StateMachine:
    def __init__(self):
        self.reset()

    def reset(self):
        self.state = UTF8State.START
        self.pending_bytes = 0
        self.expected_total = 0
        self.valid_chars = 0
        self.invalid_bytes = 0
        self.total_bytes = 0
        self.char_boundaries: List[int] = []
        self.current_char_bytes: List[int] = []

    def _is_continuation(self, byte: int) -> bool:
        return 0x80 <= byte <= 0xBF

    def _complete_char(self):
        self.valid_chars += 1
        self.char_boundaries.append(self.total_bytes)
        self.state = UTF8State.START
        self.pending_bytes = 0
        self.expected_total = 0
        self.current_char_bytes = []

    def _mark_invalid(self, byte: int):
        self.invalid_bytes += 1
        self.current_char_bytes = []
        self.state = UTF8State.START
        self.pending_bytes = 0
        self.expected_total = 0

    def process_byte(self, byte: int):
        self.total_bytes += 1

        if self.state == UTF8State.START:
            if byte <= 0x7F:
                self.current_char_bytes = [byte]
                self._complete_char()
            elif byte <= 0xBF:
                self._mark_invalid(byte)
            elif byte <= 0xDF:
                if byte <= 0xC1:
                    self._mark_invalid(byte)
                else:
                    self.state = UTF8State.EXPECT_1
                    self.pending_bytes = 1
                    self.expected_total = 2
                    self.current_char_bytes = [byte]
            elif byte <= 0xEF:
                self.state = UTF8State.EXPECT_2_AFTER_1
                self.pending_bytes = 2
                self.expected_total = 3
                self.current_char_bytes = [byte]
            elif byte <= 0xF7:
                self.state = UTF8State.EXPECT_3_AFTER_1
                self.pending_bytes = 3
                self.expected_total = 4
                self.current_char_bytes = [byte]
            else:
                self._mark_invalid(byte)

        elif self.state == UTF8State.EXPECT_1:
            if self._is_continuation(byte):
                self.current_char_bytes.append(byte)
                self._complete_char()
            else:
                self.invalid_bytes += 1
                self.current_char_bytes = []
                self.state = UTF8State.START
                self.pending_bytes = 0
                self.expected_total = 0
                self.total_bytes -= 1
                self.process_byte(byte)

        elif self.state == UTF8State.EXPECT_2_AFTER_1:
            lead = self.current_char_bytes[0]
            if self._is_continuation(byte):
                if lead == 0xE0 and byte < 0xA0:
                    self._mark_invalid(byte)
                elif lead == 0xED and byte >= 0xA0:
                    self._mark_invalid(byte)
                else:
                    self.current_char_bytes.append(byte)
                    self.state = UTF8State.EXPECT_2
                    self.pending_bytes = 1
            else:
                self.invalid_bytes += 1
                self.current_char_bytes = []
                self.state = UTF8State.START
                self.pending_bytes = 0
                self.expected_total = 0
                self.total_bytes -= 1
                self.process_byte(byte)

        elif self.state == UTF8State.EXPECT_2:
            if self._is_continuation(byte):
                self.current_char_bytes.append(byte)
                self._complete_char()
            else:
                self.invalid_bytes += 2
                self.current_char_bytes = []
                self.state = UTF8State.START
                self.pending_bytes = 0
                self.expected_total = 0
                self.total_bytes -= 1
                self.process_byte(byte)

        elif self.state == UTF8State.EXPECT_3_AFTER_1:
            lead = self.current_char_bytes[0]
            if self._is_continuation(byte):
                if lead == 0xF0 and byte < 0x90:
                    self._mark_invalid(byte)
                elif lead == 0xF4 and byte >= 0x90:
                    self._mark_invalid(byte)
                elif lead > 0xF4:
                    self._mark_invalid(byte)
                else:
                    self.current_char_bytes.append(byte)
                    self.state = UTF8State.EXPECT_3_AFTER_2
                    self.pending_bytes = 2
            else:
                self.invalid_bytes += 1
                self.current_char_bytes = []
                self.state = UTF8State.START
                self.pending_bytes = 0
                self.expected_total = 0
                self.total_bytes -= 1
                self.process_byte(byte)

        elif self.state == UTF8State.EXPECT_3_AFTER_2:
            if self._is_continuation(byte):
                self.current_char_bytes.append(byte)
                self.state = UTF8State.EXPECT_3
                self.pending_bytes = 1
            else:
                self.invalid_bytes += 2
                self.current_char_bytes = []
                self.state = UTF8State.START
                self.pending_bytes = 0
                self.expected_total = 0
                self.total_bytes -= 1
                self.process_byte(byte)

        elif self.state == UTF8State.EXPECT_3:
            if self._is_continuation(byte):
                self.current_char_bytes.append(byte)
                self._complete_char()
            else:
                self.invalid_bytes += 3
                self.current_char_bytes = []
                self.state = UTF8State.START
                self.pending_bytes = 0
                self.expected_total = 0
                self.total_bytes -= 1
                self.process_byte(byte)

    def process_bytes(self, data: bytes) -> UTF8Analysis:
        self.reset()
        for byte in data:
            self.process_byte(byte)

        return UTF8Analysis(
            valid_chars=self.valid_chars,
            invalid_bytes=self.invalid_bytes,
            total_bytes=self.total_bytes,
            final_state=self.state,
            pending_bytes=self.pending_bytes,
            expected_total=self.expected_total,
            char_boundaries=self.char_boundaries.copy(),
        )


@dataclass
class DualScore:
    partial_score: float
    binary_strict: float
    binary_soft: float
    valid_chars: int
    invalid_bytes: int
    total_bytes: int
    is_complete: bool
    is_valid_prefix: bool
    incomplete_progress: float
    final_state: str


def compute_dual_score(byte_data: bytes) -> DualScore:
    sm = UTF8StateMachine()
    analysis = sm.process_bytes(byte_data)

    if analysis.total_bytes == 0:
        return DualScore(
            partial_score=1.0,
            binary_strict=1.0,
            binary_soft=1.0,
            valid_chars=0,
            invalid_bytes=0,
            total_bytes=0,
            is_complete=True,
            is_valid_prefix=True,
            incomplete_progress=1.0,
            final_state="START",
        )

    valid_char_bytes = analysis.char_boundaries[-1] if analysis.char_boundaries else 0

    incomplete_credit = 0.0
    if analysis.final_state not in (UTF8State.START, UTF8State.INVALID):
        bytes_in_incomplete = analysis.expected_total - analysis.pending_bytes
        incomplete_credit = bytes_in_incomplete * analysis.incomplete_progress

    partial_score = (valid_char_bytes + incomplete_credit) / analysis.total_bytes
    binary_strict = 1.0 if (analysis.is_complete and analysis.invalid_bytes == 0) else 0.0
    binary_soft = valid_char_bytes / analysis.total_bytes

    return DualScore(
        partial_score=partial_score,
        binary_strict=binary_strict,
        binary_soft=binary_soft,
        valid_chars=analysis.valid_chars,
        invalid_bytes=analysis.invalid_bytes,
        total_bytes=analysis.total_bytes,
        is_complete=analysis.is_complete,
        is_valid_prefix=analysis.is_valid_prefix,
        incomplete_progress=analysis.incomplete_progress,
        final_state=analysis.final_state.name,
    )

File: ts/src/dfa.ts
/**
 * UTF-8 DFA state machine and dual scoring.
 *
 * Provides V_partial, V_binary_strict, V_binary_soft scoring
 * matching the Python bytecanary implementation exactly.
 */

export enum UTF8State {
  START,
  EXPECT_1,
  EXPECT_2,
  EXPECT_3,
  EXPECT_2_AFTER_1,
  EXPECT_3_AFTER_1,
  EXPECT_3_AFTER_2,
  INVALID,
}

const STATE_PROGRESS: Record<UTF8State, number> = {
  [UTF8State.START]: 1.0,
  [UTF8State.EXPECT_1]: 0.5,
  [UTF8State.EXPECT_2]: 0.333,
  [UTF8State.EXPECT_2_AFTER_1]: 0.667,
  [UTF8State.EXPECT_3]: 0.25,
  [UTF8State.EXPECT_3_AFTER_1]: 0.5,
  [UTF8State.EXPECT_3_AFTER_2]: 0.75,
  [UTF8State.INVALID]: 0.0,
};

export interface UTF8Analysis {
  validChars: number;
  invalidBytes: number;
  totalBytes: number;
  finalState: UTF8State;
  pendingBytes: number;
  expectedTotal: number;
  charBoundaries: number[];
  isComplete: boolean;
  isValidPrefix: boolean;
  incompleteProgress: number;
}

export class UTF8StateMachine {
  private state = UTF8State.START;
  private pendingBytes = 0;
  private expectedTotal = 0;
  private validChars = 0;
  private invalidBytes = 0;
  private totalBytes = 0;
  private charBoundaries: number[] = [];
  private currentCharBytes: number[] = [];

  reset(): void {
    this.state = UTF8State.START;
    this.pendingBytes = 0;
    this.expectedTotal = 0;
    this.validChars = 0;
    this.invalidBytes = 0;
    this.totalBytes = 0;
    this.charBoundaries = [];
    this.currentCharBytes = [];
  }

  private isContinuation(byte: number): boolean {
    return byte >= 0x80 && byte <= 0xbf;
  }

  private completeChar(): void {
    this.validChars++;
    this.charBoundaries.push(this.totalBytes);
    this.state = UTF8State.START;
    this.pendingBytes = 0;
    this.expectedTotal = 0;
    this.currentCharBytes = [];
  }

  private markInvalid(_byte: number): void {
    this.invalidBytes++;
    this.currentCharBytes = [];
    this.state = UTF8State.START;
    this.pendingBytes = 0;
    this.expectedTotal = 0;
  }

  processByte(byte: number): void {
    this.totalBytes++;

    if (this.state === UTF8State.START) {
      if (byte <= 0x7f) {
        this.currentCharBytes = [byte];
        this.completeChar();
      } else if (byte <= 0xbf) {
        this.markInvalid(byte);
      } else if (byte <= 0xdf) {
        if (byte <= 0xc1) {
          this.markInvalid(byte);
        } else {
          this.state = UTF8State.EXPECT_1;
          this.pendingBytes = 1;
          this.expectedTotal = 2;
          this.currentCharBytes = [byte];
        }
      } else if (byte <= 0xef) {
        this.state = UTF8State.EXPECT_2_AFTER_1;
        this.pendingBytes = 2;
        this.expectedTotal = 3;
        this.currentCharBytes = [byte];
      } else if (byte <= 0xf7) {
        this.state = UTF8State.EXPECT_3_AFTER_1;
        this.pendingBytes = 3;
        this.expectedTotal = 4;
        this.currentCharBytes = [byte];
      } else {
        this.markInvalid(byte);
      }
    } else if (this.state === UTF8State.EXPECT_1) {
      if (this.isContinuation(byte)) {
        this.currentCharBytes.push(byte);
        this.completeChar();
      } else {
        this.invalidBytes++;
        this.currentCharBytes = [];
        this.state = UTF8State.START;
        this.pendingBytes = 0;
        this.expectedTotal = 0;
        this.totalBytes--;
        this.processByte(byte);
      }
    } else if (this.state === UTF8State.EXPECT_2_AFTER_1) {
      const lead = this.currentCharBytes[0];
      if (this.isContinuation(byte)) {
        if (lead === 0xe0 && byte < 0xa0) {
          this.markInvalid(byte);
        } else if (lead === 0xed && byte >= 0xa0) {
          this.markInvalid(byte);
        } else {
          this.currentCharBytes.push(byte);
          this.state = UTF8State.EXPECT_2;
          this.pendingBytes = 1;
        }
      } else {
        this.invalidBytes++;
        this.currentCharBytes = [];
        this.state = UTF8State.START;
        this.pendingBytes = 0;
        this.expectedTotal = 0;
        this.totalBytes--;
        this.processByte(byte);
      }
    } else if (this.state === UTF8State.EXPECT_2) {
      if (this.isContinuation(byte)) {
        this.currentCharBytes.push(byte);
        this.completeChar();
      } else {
        this.invalidBytes += 2;
        this.currentCharBytes = [];
        this.state = UTF8State.START;
        this.pendingBytes = 0;
        this.expectedTotal = 0;
        this.totalBytes--;
        this.processByte(byte);
      }
    } else if (this.state === UTF8State.EXPECT_3_AFTER_1) {
      const lead = this.currentCharBytes[0];
      if (this.isContinuation(byte)) {
        if (lead === 0xf0 && byte < 0x90) {
          this.markInvalid(byte);
        } else if (lead === 0xf4 && byte >= 0x90) {
          this.markInvalid(byte);
        } else if (lead > 0xf4) {
          this.markInvalid(byte);
        } else {
          this.currentCharBytes.push(byte);
          this.state = UTF8State.EXPECT_3_AFTER_2;
          this.pendingBytes = 2;
        }
      } else {
        this.invalidBytes++;
        this.currentCharBytes = [];
        this.state = UTF8State.START;
        this.pendingBytes = 0;
        this.expectedTotal = 0;
        this.totalBytes--;
        this.processByte(byte);
      }
    } else if (this.state === UTF8State.EXPECT_3_AFTER_2) {
      if (this.isContinuation(byte)) {
        this.currentCharBytes.push(byte);
        this.state = UTF8State.EXPECT_3;
        this.pendingBytes = 1;
      } else {
        this.invalidBytes += 2;
        this.currentCharBytes = [];
        this.state = UTF8State.START;
        this.pendingBytes = 0;
        this.expectedTotal = 0;
        this.totalBytes--;
        this.processByte(byte);
      }
    } else if (this.state === UTF8State.EXPECT_3) {
      if (this.isContinuation(byte)) {
        this.currentCharBytes.push(byte);
        this.completeChar();
      } else {
        this.invalidBytes += 3;
        this.currentCharBytes = [];
        this.state = UTF8State.START;
        this.pendingBytes = 0;
        this.expectedTotal = 0;
        this.totalBytes--;
        this.processByte(byte);
      }
    }
  }

  processBytes(data: Uint8Array): UTF8Analysis {
    this.reset();
    for (let i = 0; i < data.length; i++) {
      this.processByte(data[i]);
    }
    return {
      validChars: this.validChars,
      invalidBytes: this.invalidBytes,
      totalBytes: this.totalBytes,
      finalState: this.state,
      pendingBytes: this.pendingBytes,
      expectedTotal: this.expectedTotal,
      charBoundaries: [...this.charBoundaries],
      isComplete: this.state === UTF8State.START,
      isValidPrefix:
        this.state !== UTF8State.INVALID && this.invalidBytes === 0,
      incompleteProgress: STATE_PROGRESS[this.state] ?? 0.0,
    };
  }
}

export interface DualScore {
  partialScore: number;
  binaryStrict: number;
  binarySoft: number;
  validChars: number;
  invalidBytes: number;
  totalBytes: number;
  isComplete: boolean;
  isValidPrefix: boolean;
  incompleteProgress: number;
  finalState: string;
}

export function computeDualScore(data: Uint8Array): DualScore {
  const sm = new UTF8StateMachine();
  const analysis = sm.processBytes(data);

  if (analysis.totalBytes === 0) {
    return {
      partialScore: 1.0,
      binaryStrict: 1.0,
      binarySoft: 1.0,
      validChars: 0,
      invalidBytes: 0,
      totalBytes: 0,
      isComplete: true,
      isValidPrefix: true,
      incompleteProgress: 1.0,
      finalState: "START",
    };
  }

  const validCharBytes =
    analysis.charBoundaries.length > 0
      ? analysis.charBoundaries[analysis.charBoundaries.length - 1]
      : 0;

  let incompleteCredit = 0.0;
  if (
    analysis.finalState !== UTF8State.START &&
    analysis.finalState !== UTF8State.INVALID
  ) {
    const bytesInIncomplete =
      analysis.expectedTotal - analysis.pendingBytes;
    incompleteCredit = bytesInIncomplete * analysis.incompleteProgress;
  }

  const partialScore =
    (validCharBytes + incompleteCredit) / analysis.totalBytes;
  const binaryStrict =
    analysis.isComplete && analysis.invalidBytes === 0 ? 1.0 : 0.0;
  const binarySoft = validCharBytes / analysis.totalBytes;

  return {
    partialScore,
    binaryStrict,
    binarySoft,
    validChars: analysis.validChars,
    invalidBytes: analysis.invalidBytes,
    totalBytes: analysis.totalBytes,
    isComplete: analysis.isComplete,
    isValidPrefix: analysis.isValidPrefix,
    incompleteProgress: analysis.incompleteProgress,
    finalState: UTF8State[analysis.finalState],
  };
}
